### DISCUSSION TITLE
Breaking Changes Allowed?

### DISCUSSION THREAD

**Discussion by ds5678** [xamarin/TorchSharp#1378] [Q&A]
Hello! I'm interesting in contributing. I'm looking for some clarification about what kinds of breaking changes are allowed, if any.

## Obsolete Members

I noticed some of these while I was browsing the codebase. Do you mind if I do a pass and remove them all?

## Replace `T[]` with `{ReadOnly}Span<T>`

There are many places where arrays could be replaced with spans. This would reduce allocations and improve performance.

Adding span overloads to avoid breaking changes would significantly bloat the library.

## Bump to latest .NET

Precedent: https://github.com/dotnet/ClangSharp

Exclusively targeting the latest .NET version would enable us to take advantage of all the recent improvements and features. For example, `Tensor` could benefit from the generic math interfaces.

I assume the answer is no because I noticed that [Microsoft.ML.TorchSharp](https://www.nuget.org/packages/Microsoft.ML.TorchSharp/) uses .NET Standard, but it's worth discussing.

**Comment by yueyinqiu** [comment]
About the dotnet version, we'd like to make TorchSharp be able to work on all the dotnet platforms that haven't lose support, so it's `dotnet 6` and `netstandard2.0`.

And other breaking changes... Actually I've made some, so it used to be allowed, but I'm not sure for now... API compatibility check has been added in July. It's not intentionally done by the repo's manager, I think, so we still haven't find out the common routine to deal with the breaking changes.

By the way, if spans could be added as overloads, it won't be regarded as a breaking change, so that's fine. And those obsolete stuffs, maybe we could remove them all in a future "refactoring" version.

**Comment by asieradzk** [comment]
> Hello! I'm interesting in contributing. I'm looking for some clarification about what kinds of breaking changes are allowed, if any.
> ## Obsolete Members
> 
> I noticed some of these while I was browsing the codebase. Do you mind if I do a pass and remove them all?
> ## Replace `T[]` with `{ReadOnly}Span<T>`
> 
> There are many places where arrays could be replaced with spans. This would reduce allocations and improve performance.
> 
> Adding span overloads to avoid breaking changes would significantly bloat the library.
> ## Bump to latest .NET
> 
> Precedent: https://github.com/dotnet/ClangSharp
> 
> Exclusively targeting the latest .NET version would enable us to take advantage of all the recent improvements and features. For example, `Tensor` could benefit from the generic math interfaces.
> 
> I assume the answer is no because I noticed that [Microsoft.ML.TorchSharp](https://www.nuget.org/packages/Microsoft.ML.TorchSharp/) uses .NET Standard, but it's worth discussing.

If you make a fork that only targets .net 8 or 9 I am 100% moving right away. Recently I've had many disappointments with standard 2.0 and its not worth maintaining it for me.

**Reply by ds5678** [reply]
I'm not interested in doing a hard fork of the project, other than maybe just some modifications for myself. Specifically, I'm not interested in the maintenance and completeness required for a hard fork.

**Comment by NiklasGustafsson** [comment]
Regarding targeted versions of .NET -- we are completely committed to supporting any version of .NET that is still supported by the .NET team. That includes .NET FX 4.7.2 and 4.8.

I believe .NET 6 is out of support, and if so, we can target .NET 7 and later.

**Reply by ds5678** [reply]
Thank you for the response. .NET 7 has been out of support since May 14 of this year, so I assume you mean .NET 8?

**Reply by NiklasGustafsson** [reply]
Sorry, yes. I got the version numbers confused. 

.NET FX is a more pressing constraint on the API, though.
